Framework/Algorithms/src/GetDetectorOffsets.cpp

```diff
@@ -107,16 +129,46 @@ void GetDetectorOffsets::exec() {
   const detid2index_map pixel_to_wi =
       maskWS->getDetectorIDToWorkspaceIndexMap(true);
 
+  // fit each peak twice?
+  bool fit_peak_twice = getProperty("FitEachPeakTwice");
+  double minimum_peak_height = getProperty("MinimumPeakHeight");
+
+  // output fitting result?
+  bool output_fit_result = getProperty("OutputFitResult");
+  TableWorkspace_sptr fit_result_table(0);
+  if (output_fit_result)
+    fit_result_table = GenerateFitResultTable();
+
   // Fit all the spectra with a gaussian
   Progress prog(this, 0.0, 1.0, nspec);
   auto &spectrumInfo = maskWS->mutableSpectrumInfo();
   PARALLEL_FOR_IF(Kernel::threadSafe(*inputW))
   for (int64_t wi = 0; wi < nspec; ++wi) {
     PARALLEL_START_INTERUPT_REGION
     // Fit the peak
-    double offset = fitSpectra(wi, isAbsolute);
+    GetDetectorsOffset::PeakLinearFunction fit_result;
+    double offset =
+        fitSpectra(wi, isAbsolute, m_Xmin, m_Xmax, fit_result, false);
+
+    // fit peak for the second time
+    bool mask_it(false);
+    if (fit_peak_twice) {
+      double offset2 = fitPeakSecondTime(wi, isAbsolute, minimum_peak_height,
+                                         fit_result, mask_it);
+      if (mask_it) {
+        if (offset2 < -1.9)
+          g_log.debug() << "ws-index " << wi << " has NaN.";
+        else
+          g_log.debug() << "ws-index " << wi << " has either too-low peak ("
+                        << fit_result.height << ") or too-narrow peak ("
+                        << fit_result.sigma << ")\n";
+      }
+
+      offset = offset2;
+    }
+
     double mask = 0.0;
-    if (std::abs(offset) > m_maxOffset) {
+    if (mask_it || std::abs(offset) > m_maxOffset) {
```

Comment:
Fold this into the logic of lines 154-168.

Reply:
change here is not trivial.  need to look twice.